File: backend/src/test/java/com/skewwhiffy/auraltester/controller/KeySignatureControllerTest.java
package com.skewwhiffy.auraltester.controller;

/*
@ExtendWith(MockitoExtension::class)
class KeySignatureControllerTest {
  @Mock
  private InternalNotationFactory internalNotationFactory;

  @Mock
  private AbcService abcService;

  @InjectMocks
  private KeySignatureController keySignatureController;
  private Clef clef;
  private String clefString;
  private String keySignature;
  private AbsoluteNote keySignatureNote;
  private String abc;

  @BeforeEach
  fun `set up`() {
    clef = TestData.random.clef
    clefString = TestData.random.string
    keySignature = TestData.random.string
    keySignatureNote = TestData.random.absoluteNote
    abc = TestData.random.string
  }

  @Test
  fun `responds correctly`() {
    `when`(internalNotationFactory.clef(clefString)).thenReturn(clef)
    `when`(internalNotationFactory.getNote(keySignature)).thenReturn(keySignatureNote)
    `when`(abcService.getAbc(clef, Key(keySignatureNote.note))).thenReturn(abc.let {
      object : AbcProvider {
        override val abc: String = it
      }
    })

    val actual = keySignatureController.get(clefString, keySignature)

    assertThat(actual.abc).isEqualTo(abc)
  }
}

 */

File: backend/src/test/kotlin/com/skewwhiffy/auraltester/controller/KeySignatureControllerTest.kt
package com.skewwhiffy.auraltester.controller

import com.skewwhiffy.auraltester.notation.factory.InternalNotationFactory
import com.skewwhiffy.auraltester.notation.model.abc.AbcProvider
import com.skewwhiffy.auraltester.notation.model.clef.Clef
import com.skewwhiffy.auraltester.notation.model.key.Key
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import com.skewwhiffy.auraltester.service.AbcService
import com.skewwhiffy.auraltester.test.util.TestData
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith


@ExtendWith(MockKExtension::class)
class KeySignatureControllerTest {
    @MockK
    private lateinit var internalNotationFactory: InternalNotationFactory

    @MockK
    private lateinit var abcService: AbcService

    @InjectMockKs
    private lateinit var keySignatureController: KeySignatureController
    private lateinit var clef: Clef
    private lateinit var clefString: String
    private lateinit var keySignature: String
    private lateinit var keySignatureNote: AbsoluteNote
    private lateinit var abc: String

    @BeforeEach
    fun setUp() {
        clef = TestData.random.clef
        clefString = TestData.random.string
        keySignature = TestData.random.string
        keySignatureNote = TestData.random.absoluteNote
        abc = TestData.random.string
    }

    @Test
    fun respondsCorrectly() {
        every { internalNotationFactory.clef(clefString) } returns clef
        every { internalNotationFactory.getNote(keySignature) } returns keySignatureNote
        val abcResult = abc
        every { abcService.getAbc(clef, Key.major(keySignatureNote.note)) } returns object : AbcProvider {
            override val abc = abcResult
        }
        val actual = keySignatureController.get(clefString, keySignature)

        assertThat(actual.abc).isEqualTo(abc)
    }
}
